File: SDK Submissions/chimoney_rust_sdk/src/account/account.rs
use crate::core::api_client::APIClient;
use serde_json::json;

/// Get transaction details by issueID.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `issue_id` - The issueID of the transaction to be retrieved.
/// * `sub_account` - The sub-account associated with the transaction.
///
/// # Returns
///
/// A `Result` containing a JSON string with the transaction details if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::get_transaction_details;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let issue_id = "123456789";
/// let sub_account = "1234567";
///
/// match get_transaction_details(&api_client, issue_id, sub_account).await {
///    Ok(result) => println!("Transaction Details: {}", result),
///   Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn get_transaction_details(
    api_client: &APIClient,
    issue_id: &str,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/issue-id-transactions";

    let mut query = String::new();
    query.push_str("?issueID=");
    query.push_str(issue_id);

    let body = json!({
        "subAccount": sub_account
    });

    let res = api_client
        .post(path, &body.to_string(), Some(&query))
        .await?;
    Ok(res)
}

/// Get Public profile of a Chimoney User.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `sub_account` - The sub-account associated with the user.
/// * `user_id` - The user ID of the user.
/// * `link_code` - The link code of the user.
///
/// # Returns
///
/// A `Result` containing a JSON string with the user's public profile if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney_rust_sdk::core::api_client::APIClient;
/// use chimoney_rust_sdk::account::account::get_public_profile;
///
/// # #[tokio::main]
/// # fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let sub_account = "1234567";
/// let user_id = "123456789";
/// let link_code = "123456789";
///
/// match get_public_profile(&api_client, sub_account, user_id, link_code).await {
///    Ok(result) => println!("Public Profile: {}", result),
///  Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn get_public_profile(
    api_client: &APIClient,
    sub_account: &str,
    user_id: &str,
    link_code: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/public-profile";

    let body = json!({
        "subAccount": sub_account,
        "userID": user_id,
        "linkCode": link_code
    });

    let res = api_client.post(path, &body.to_string(), None).await?;
    Ok(res)
}

/// Get all transactions by account.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `sub_account` - The sub-account associated with the transactions.
///
/// # Returns
///
/// A `Result` containing a JSON string with the transactions if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::get_transactions_by_account;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let sub_account = "1234567";
///
/// match get_transactions_by_account(&api_client, sub_account).await {
///   Ok(result) => println!("Transactions: {}", result),
///  Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn get_transactions_by_account(
    api_client: &APIClient,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/transactions";

    let body = json!({
        "subAccount": sub_account
    });

    let res = api_client.post(path, &body.to_string(), None).await?;
    Ok(res)
}

/// Get single transaction detail.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `id` - The ID of the transaction to be retrieved.
/// * `sub_account` - The sub-account associated with the transaction.
///
/// # Returns
///
/// A `Result` containing a JSON string with the transaction details if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::get_transaction_detail;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let id = "123456789";
/// let sub_account = "1234567";
///
/// match get_transaction_detail(&api_client, id, sub_account).await {
///  Ok(result) => println!("Transaction Detail: {}", result),
/// Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn get_transaction_detail(
    api_client: &APIClient,
    id: &str,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/transaction";

    let mut query = String::new();
    query.push_str("?id=");
    query.push_str(id);

    let body = json!({
        "subAccount": sub_account,
    });

    let res = api_client.post(path, &body.to_string(), None).await?;
    Ok(res)
}

/// Account transfer.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `receiver` - The receiver of the transfer.
/// * `value_in_usd` - The value of the transfer in USD.
/// * `wallet` - The wallet associated with the transfer.
/// * `sub_account` - The sub-account associated with the transfer.
///
/// # Returns
///
/// A `Result` containing a JSON string with the result of the transfer if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::account_transfer;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let receiver = "123456789";
/// let value_in_usd = 100;
/// let wallet = "123456789";
/// let sub_account = "1234567";
///
/// match account_transfer(&api_client, receiver, &value_in_usd, wallet, sub_account).await {
///    Ok(result) => println!("Transfer Result: {}", result),
///  Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
/// # Notes
///
/// The `value_in_usd` argument must be an integer.
///
pub async fn account_transfer(
    api_client: &APIClient,
    receiver: &str,
    value_in_usd: &i32,
    wallet: &str,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/transfer";

    let body = json!({
        "receiver": receiver,
        "valueInUSD": value_in_usd,
        "wallet": wallet,
        "subAccount": sub_account,
    });

    let res = api_client.post(path, &body.to_string(), None).await?;
    Ok(res)
}

/// Delete an unpaid transaction.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `chi_ref` - The CHI reference of the unpaid transaction to be deleted.
/// * `sub_account` - The sub-account associated with the unpaid transaction.
///
/// # Returns
///
/// A `Result` containing a JSON string with the result of the deletion if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::delete_unpaid_transaction;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let chi_ref = "123456789";
/// let sub_account = "1234567";
///
/// match delete_unpaid_transaction(&api_client, chi_ref, sub_account).await {
///   Ok(result) => println!("Deletion Result: {}", result),
/// Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn delete_unpaid_transaction(
    api_client: &APIClient,
    chi_ref: &str,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/accounts/delete-unpaid-transaction";

    let mut query = String::new();
    query.push_str("?chiRef=");
    query.push_str(chi_ref);
    query.push_str("&subAccount=");
    query.push_str(sub_account);

    let res = api_client.delete(path, Some(&query)).await?;
    Ok(res)
}

/// Retrieve unpaid transactions for a specific sub-account.
///
/// # Arguments
///
/// * `api_client` - An instance of the `APIClient` used to make API requests.
/// * `sub_account` - The sub-account for which unpaid transactions should be retrieved.
///
/// # Returns
///
/// A `Result` containing a JSON string with the unpaid transactions if successful,
/// or an error if the request fails.
///
/// # Examples
///
/// ```rust
/// use chimoney::core::api_client::APIClient;
/// use chimoney::account::account::get_unpaid_transactions;
///
/// # #[tokio::main]
/// # async fn main() {
/// let api_client = APIClient::new("https://api.example.com");
/// let sub_account = "1234567";
///
/// match get_unpaid_transactions(&api_client, sub_account).await {
///    Ok(result) => println!("Unpaid Transactions: {}", result),
/// Err(err) => eprintln!("Error: {}", err),
/// }
/// # }
/// ```
///
pub async fn get_unpaid_transaction(
    api_client: &APIClient,
    sub_account: &str,
) -> Result<String, Box<dyn std::error::Error>> {
    let path = "/v0.2/beneficiary";

    let mut query = String::new();
    query.push_str("?subAccount=");
    query.push_str(sub_account);

    let res = api_client.get(path, Some(&query)).await?;
    Ok(res)
}

File: SDK Submissions/Chimoney-Java/chimoney/src/main/java/io/chimoney/chimoney/Account.java
package io.chimoney.chimoney;

import java.net.http.HttpResponse;
import java.util.List;
import java.util.Map;

import org.json.JSONObject;

/**
 * The Account class is a wrapper for the Account module of the Chimoney API.
 * 
 */
public class Account extends Base {

	/**
	 * Constructs a new Account object with the specified Chimoney instance.
	 * 
	 * @param chimoney the Chimoney instance associated with the account
	 */
	public Account(Chimoney chimoney) {
		super(chimoney);
	}

	/**
	 * Retrieves a list of transactions by issue ID.
	 *
	 * @param issueID the issueID of the transaction
	 * @return a list of transactions
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public List<Object> getTransactionsByIssueID(String issueID) throws ChimoneyException {
		return getTransactionsByIssueID(issueID, null);
	}

	/**
	 * Retrieves a list of transactions by issue ID and sub-account.
	 *
	 * @param issueID    the issue ID of the transactions to retrieve
	 * @param subAccount the sub-account associated with the transactions
	 * @return a list of transactions as objects
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public List<Object> getTransactionsByIssueID(String issueID, String subAccount) throws ChimoneyException {
		JSONObject jsonParams = new JSONObject();

		if (subAccount != null)
			jsonParams.put("subAccount", subAccount);

		HttpResponse<String> response = handlePOSTRequest("accounts/issue-id-transactions?issueID=" + issueID,
				jsonParams);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONArray("data").toList();
	}

	/**
	 * Retrieves the profile information for a user.
	 * Specify either userID or linkCode. If both are specified, userID will be
	 * used.
	 *
	 * @param userID   the ID of the user
	 * @param linkCode the link code associated with the user
	 * @return a map containing the profile information
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> getProfile(String userID, String linkCode) throws ChimoneyException {
		return getProfile(userID, linkCode, null);
	}

	/**
	 * Retrieves the profile information for a user and sub-account.
	 * Specify either userID or linkCode. If both are specified, userID will be
	 * used.
	 *
	 * @param userID     the ID of the user
	 * @param linkCode   the link code associated with the user
	 * @param subAccount the sub-account associated with the user
	 * @return a map containing the profile information
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> getProfile(String userID, String linkCode, String subAccount) throws ChimoneyException {
		JSONObject jsonParams = new JSONObject();
		if (userID != null)
			jsonParams.put("userID", userID);
		else if (linkCode != null)
			jsonParams.put("linkCode", linkCode);
		else
			throw new ChimoneyException("Either userID or linkCode must be specified.");

		if (subAccount != null)
			jsonParams.put("subAccount", subAccount);

		HttpResponse<String> response = handlePOSTRequest("accounts/public-profile", jsonParams);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONObject("data").toMap();
	}

	/**
	 * Retrieves a list of transactions for this account.
	 *
	 * @return a list of transactions for this account
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public List<Object> getTransactionsByAccount() throws ChimoneyException {
		return getTransactionsByAccount(null);
	}

	/**
	 * Retrieves a list of transactions for a specific sub-account.
	 *
	 * @param subAccount the subAccount for which to retrieve transactions
	 * @return a list of transactions as objects
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public List<Object> getTransactionsByAccount(String subAccount) throws ChimoneyException {
		JSONObject jsonParams = new JSONObject();
		jsonParams.put("subAccount", subAccount);

		HttpResponse<String> response = handlePOSTRequest("accounts/transactions", jsonParams);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONArray("data").toList();
	}

	/**
	 * Retrieves a transaction from this account by the transaction ID.
	 *
	 * @param id The ID of the transaction to retrieve.
	 * @return A map containing the transaction data.
	 * @throws ChimoneyException If an error is returned by the Chimoney API
	 */
	public Map<String, Object> getTransaction(String id) throws ChimoneyException {
		return getTransaction(id, null);
	}

	/**
	 * Retrieves a transaction from this account by the transaction ID.
	 *
	 * @param id         The ID of the transaction to retrieve.
	 * @param subAccount The sub-account associated with the transaction.
	 * @return A map containing the transaction data.
	 * @throws ChimoneyException If an error is returned by the Chimoney API
	 */
	public Map<String, Object> getTransaction(String id, String subAccount) throws ChimoneyException {
		JSONObject jsonParams = new JSONObject();

		if (subAccount != null)
			jsonParams.put("subAccount", subAccount);

		HttpResponse<String> response = handlePOSTRequest("accounts/transaction?id=" + id, jsonParams);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONObject("data").toMap();
	}

	/**
	 * Transfers funds from this account to the specified receiver.
	 *
	 * @param receiver   the ID of the recipient of the funds
	 * @param valueInUSD the amount to transfer in USD
	 * @return a map containing the details of the transfer
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> accountTransfer(String receiver, int valueInUSD) throws ChimoneyException {
		return accountTransfer(receiver, valueInUSD, null, null);
	}

	/**
	 * Transfers funds from this account to the specified receiver.
	 *
	 * @param receiver   the ID of the recipient of the funds
	 * @param valueInUSD the amount to transfer in USD
	 * @param wallet     the wallet type to use for the transfer
	 * @return a map containing the details of the transfer
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> accountTransfer(String receiver, int valueInUSD, Wallet.Types wallet)
			throws ChimoneyException {
		return accountTransfer(receiver, valueInUSD, wallet, null);
	}

	/**
	 * Transfers funds from this account to the specified receiver.
	 *
	 * @param receiver   the ID of the recipient of the funds
	 * @param valueInUSD the amount to transfer in USD
	 * @param wallet     the wallet type to use for the transfer
	 * @param subAccount the sub-account to use for the transfer
	 * @return a map containing the response data from the transfer operation
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> accountTransfer(String receiver, int valueInUSD, Wallet.Types wallet, String subAccount)
			throws ChimoneyException {
		JSONObject jsonParams = new JSONObject();
		jsonParams.put("receiver", receiver);
		jsonParams.put("valueInUSD", valueInUSD);

		if (wallet != null)
			jsonParams.put("wallet", wallet.toString());

		if (subAccount != null)
			jsonParams.put("subAccount", subAccount);

		HttpResponse<String> response = handlePOSTRequest("accounts/transfer", jsonParams);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONObject("data").toMap();
	}

	/**
	 * Deletes an unpaid transaction with the specified chiRef.
	 *
	 * @param chiRef the reference of the transaction to be deleted
	 * @return a map containing the result of the deletion operation
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> deleteUnpaidTransaction(String chiRef) throws ChimoneyException {
		return deleteUnpaidTransaction(chiRef, null);
	}

	/**
	 * Deletes an unpaid transaction with the specified chiRef.
	 *
	 * @param chiRef     the reference of the transaction to be deleted
	 * @param subAccount the subAccount associated with the transaction
	 * @return a map containing the data of the deleted transaction
	 * @throws ChimoneyException if an error is returned by the Chimoney API
	 */
	public Map<String, Object> deleteUnpaidTransaction(String chiRef, String subAccount)
			throws ChimoneyException {
		String ext = subAccount != null ? "&subAccount=" + subAccount : "";
		HttpResponse<String> response = handleDELETERequest(
				"accounts/delete-unpaid-transaction?chiRef=" + chiRef + ext);
		JSONObject jo = parseJSONData(response);

		return jo.getJSONObject("data").toMap();
	}
}
